File: backend/app/schemas/principal.py
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr, ConfigDict


# ---------------------------------------------------------------------------
# School overview
# ---------------------------------------------------------------------------
class SchoolOut(BaseModel):
    id: UUID
    name: str
    address: Optional[str] = None
    phone: Optional[str] = None
    website: Optional[str] = None
    created_at: datetime
    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Student management
# ---------------------------------------------------------------------------
class StudentCreateByPrincipal(BaseModel):
    email: EmailStr
    password: str
    full_name: str
    student_id_number: Optional[str] = None
    grade_level: Optional[int] = None
    class_id: Optional[UUID] = None  # optionally enroll immediately


class StudentUpdateByPrincipal(BaseModel):
    full_name: Optional[str] = None
    student_id_number: Optional[str] = None
    grade_level: Optional[int] = None
    is_active: Optional[bool] = None


class StudentManageOut(BaseModel):
    user_id: UUID
    student_id: UUID
    email: str
    full_name: Optional[str] = None
    is_active: bool
    student_id_number: Optional[str] = None
    grade_level: Optional[int] = None
    class_id: Optional[UUID] = None
    class_name: Optional[str] = None


# ---------------------------------------------------------------------------
# Teacher management
# ---------------------------------------------------------------------------
class TeacherCreateByPrincipal(BaseModel):
    email: EmailStr
    password: str
    full_name: str
    employee_id: Optional[str] = None
    specialization: Optional[str] = None


class TeacherUpdateByPrincipal(BaseModel):
    full_name: Optional[str] = None
    employee_id: Optional[str] = None
    specialization: Optional[str] = None
    is_active: Optional[bool] = None


class TeacherManageOut(BaseModel):
    user_id: UUID
    teacher_id: UUID
    email: str
    full_name: Optional[str] = None
    is_active: bool
    employee_id: Optional[str] = None
    specialization: Optional[str] = None
    subjects_count: int = 0


# ---------------------------------------------------------------------------
# Attendance monitoring
# ---------------------------------------------------------------------------
class ClassAttendanceSummary(BaseModel):
    class_id: UUID
    class_name: str
    present: int
    absent: int
    late: int
    total_marked: int


class AttendanceOverviewOut(BaseModel):
    date: date
    present: int
    absent: int
    late: int
    total_marked: int
    by_class: List[ClassAttendanceSummary]


# ---------------------------------------------------------------------------
# Homework / assignment monitoring
# ---------------------------------------------------------------------------
class HomeworkOverviewItem(BaseModel):
    id: UUID
    title: str
    class_id: UUID
    class_name: str
    due_date: Optional[datetime] = None
    submitted: int
    total_students: int
    completion_rate: Optional[float] = None


# ---------------------------------------------------------------------------
# Exam / results monitoring
# ---------------------------------------------------------------------------
class ExamOverviewItem(BaseModel):
    id: UUID
    title: str
    subject_id: UUID
    subject_name: str
    class_id: Optional[UUID] = None
    class_name: Optional[str] = None
    date: Optional[datetime] = None
    results_recorded: int
    average_score: Optional[float] = None

File: frontend/src/types/principal.ts
// Mirrors backend/app/schemas/principal.py

export interface PrincipalSchool {
  id: string;
  name: string;
  address: string | null;
  phone: string | null;
  website: string | null;
  created_at: string;
}

export interface StudentManageItem {
  user_id: string;
  student_id: string;
  email: string;
  full_name: string | null;
  is_active: boolean;
  student_id_number: string | null;
  grade_level: number | null;
  class_id: string | null;
  class_name: string | null;
}

export interface StudentCreateInput {
  email: string;
  password: string;
  full_name: string;
  student_id_number?: string;
  grade_level?: number;
  class_id?: string;
}

export interface StudentUpdateInput {
  full_name?: string;
  student_id_number?: string;
  grade_level?: number;
  is_active?: boolean;
}

export interface TeacherManageItem {
  user_id: string;
  teacher_id: string;
  email: string;
  full_name: string | null;
  is_active: boolean;
  employee_id: string | null;
  specialization: string | null;
  subjects_count: number;
}

export interface TeacherCreateInput {
  email: string;
  password: string;
  full_name: string;
  employee_id?: string;
  specialization?: string;
}

export interface TeacherUpdateInput {
  full_name?: string;
  employee_id?: string;
  specialization?: string;
  is_active?: boolean;
}

export interface ClassAttendanceSummary {
  class_id: string;
  class_name: string;
  present: number;
  absent: number;
  late: number;
  total_marked: number;
}

export interface AttendanceOverview {
  date: string;
  present: number;
  absent: number;
  late: number;
  total_marked: number;
  by_class: ClassAttendanceSummary[];
}

export interface HomeworkOverviewItem {
  id: string;
  title: string;
  class_id: string;
  class_name: string;
  due_date: string | null;
  submitted: number;
  total_students: number;
  completion_rate: number | null;
}

export interface ExamOverviewItem {
  id: string;
  title: string;
  subject_id: string;
  subject_name: string;
  class_id: string | null;
  class_name: string | null;
  date: string | null;
  results_recorded: number;
  average_score: number | null;
}

// Mirrors backend/app/schemas/memory.py analytics response shapes
export interface SchoolPerformance {
  school_id: string;
  school_name: string;
  average_score: number | null;
  total_students: number;
  total_teachers: number;
  total_classes: number;
}

export interface ClassPerformance {
  class_id: string;
  class_name: string;
  student_count: number;
  average_score: number | null;
  homework_completion_rate: number | null;
  attendance_rate: number | null;
}

export interface SubjectPerformance {
  subject_id: string;
  subject_name: string;
  average_score: number | null;
  total_exams: number;
}

export interface TeacherPerformance {
  teacher_id: string;
  full_name: string | null;
  subjects_taught: string[];
  average_student_score: number | null;
}

export interface AIUsageStats {
  total_conversations: number;
  total_quiz_attempts: number;
  active_students: number;
  average_conversations_per_student: number;
}

export interface PrincipalAISummary {
  data: {
    school_performance: SchoolPerformance;
    class_performance: ClassPerformance[];
    subject_performance: SubjectPerformance[];
    teacher_performance: TeacherPerformance[];
    ai_usage: AIUsageStats;
    risk_flags: Array<Record<string, unknown>>;
  };
  ai_recommendations: string | null;
}
